### PR TITLE
Major refactoring, adding systemd state and enabled

### PR DESCRIPTION
**Pull request by jstenvall-ducksify**
Principal reason for this PR is adding capability to set the state and if the unit is enabled or not.

Units/Services can now be deactivated and removed on an individual level.

It also contains a refactoring of the file structure.

Tests needs to be added. At the moment Travis CI fails miserably. My bad.

### CODE REVIEW

File: handlers/main.yml [hunk 1]
```diff
@@ -9,6 +9,35 @@
...
+  loop_control:
+    loop_var: enable_item
+  listen: "Enable systemd units"
+  when:
+    - uninstall_item.state is not defined or uninstall_item.state is defined and uninstall_item.state != "absent"
+
+- name: Uninstall systemd units
+  become: true
+  systemd:
+    name: "{{ enable_item.dest | basename }}"
```
Comment (0x0I) [inline handlers/main.yml]: => uninstall_item.dest

File: tasks/systemd.yml [hunk 2]
```diff
@@ -0,0 +1,42 @@
+---
+# receives a structure like
+# unit_item
```
Comment (0x0I) [inline tasks/systemd.yml:3]: perhaps "unit_items" since technically more than one can be passed?

File: tasks/config.yml [hunk 3]
```diff
@@ -11,7 +11,7 @@
     backup: true
```
Comment (0x0I) [inline tasks/config.yml:11]: ..so not 100% sure I understand the directory structure refactor here. I mean, 'common' was meant to allow an easy/advertised path to supporting multiple operating/distribution system types. I can see forming a precedent that placing files in the main directory marks them as "common or main" files though seems kind of arbitrary.

File: handlers/main.yml [hunk 4]
```diff
@@ -2,19 +2,9 @@
...
   become: true
   systemd:
-    name: "{{ restart_item.dest | basename }}"
+    name: "{{ reload_item.dest | basename }}"
     daemon_reload: true
-  with_items: "{{ restart_units.results }}"
+  with_items: "{{ reload_units.results }}"
   loop_control:
-    loop_var: restart_item
+    loop_var: reloadrestart_item
```
Comment (0x0I) [inline handlers/main.yml:9]: => reload_item

File: tasks/config.yml (hunk 3 above)
Comment (jstenvall-ducksify) [reply tasks/config.yml:11]: Maybe I was to rude, I have no good experience of using sub directories. The role is not extremely complex. In similar roles that I create, I usually name the files according to the function. I'll be happy to restore the folder structure if you want to.

Question: is it common that there are sub folders in roles? I cannot find that many examples.

File: handlers/main.yml (hunk 4 above)
Comment (jstenvall-ducksify) [reply handlers/main.yml:9]: Blimey, what a stupid mistake. I'm on it.

File: tasks/systemd.yml (hunk 2 above)
Comment (jstenvall-ducksify) [reply tasks/systemd.yml:3]: The systemd tasks is actually receiving only one unit at a time, please look at the `tasks/main.yml` file.

File: handlers/main.yml (hunk 1 above)
Comment (jstenvall-ducksify) [reply handlers/main.yml]: Honestly, this is ridiculous, I think I created the PR too early. I'm on it.

File: handlers/main.yml (hunk 4 above)
Comment (jstenvall-ducksify) [reply handlers/main.yml]: fixed

File: tasks/systemd.yml (hunk 2 above)
Comment (0x0I) [reply tasks/systemd.yml:3]: Actually, based on the block in `tasks/main`, I believe a list is being passed with the *include* as is (see the current implementation which has a similar construct). Thinking you already have but try placing a `debug` at the beginning of the tasks file to investigate further.

Anyway, thinking I'd save the file restructure for another commit and go with the action/activation update.

File: tasks/config.yml (hunk 3 above)
Comment (0x0I) [reply tasks/config.yml:11]: Ha no worries and it isn't at all. Definitely has to be the simplest of all the roles at least under 0xO1. As for sub folders, I'm generally just a stickler for organization but can't say I'm familiar with any sort of official convention either way.

File: tasks/systemd.yml (hunk 2 above)
Comment (0x0I) [reply tasks/systemd.yml:3]: Hmmm..maybe not. afk atm but will double-check in bit.

Comment (jstenvall-ducksify) [reply tasks/systemd.yml]: Any further comments?

**Comment by goldpanic** [comment]
Hi, any news about having the possibility to enable created services?

**Comment by jstenvall-ducksify** [comment]
> Hi, any news about having the possibility to enable created services?

Funcitonally, the PR alright. I'm using it in "production".

It's only the tests that I cannot get my head around to fix. I just do not have the time to look at the kitchen stuff.

**Comment by 0x0I** [comment]
> > Hi, any news about having the possibility to enable created services?
> 
> Funcitonally, the PR is alright. I'm using it in "production".
> 
> It's only the tests that I cannot get my head around to fix. I just do not have the time to look at the kitchen stuff.

Hey, got it and again thanks for the PR!

To be honest though, not sure I get the add for `enable` services within this role. The initial design was meant for the role to control the rendering of **systemd unit** files and allow the end-user to manage the execution/orchestration of these unit files however he/she sees fit (whether that involves Ansible's built-in set of [systemd](https://docs.ansible.com/ansible/latest/modules/systemd_module.html) or [service](https://docs.ansible.com/ansible/latest/modules/service_module.html) manager modules, cmdline, etc).

You'll notice [here](https://github.com/0x0I/ansible-role-elasticsearch/blob/master/tasks/common/launch.yml) how this design is applied with rendering and orchestration separated.

**Comment by 0x0I** [comment]
> Hi, any news about having the possibility to enable created services?

*See comment below.*

Feels like rendering of *unit* files (the responsibility of this role) and actual orchestration of the resource described by that file (responsibility of e.g. Ansible's **service** or **systemd** module) should be separate/independent. Not to mention that an orchestration solution is already well implemented (see previously mentioned modules lol).

**Comment by jstenvall-ducksify** [comment]
> > Hi, any news about having the possibility to enable created services?
> 
> _See comment below._
> 
> Feels like rendering of _unit_ files (the responsibility of this role) and actual orchestration of the resource described by that file (responsibility of e.g. Ansible's **service** or **systemd** module) should be separate/independent. Not to mention that an orchestration solution is already well implemented (see previously mentioned modules lol).

My use case is somewhat different, I include this role in the playbook and the host_var contains the `unit_config`. Thats it, there is no role, no logic at all. I only want to add the unit files and activate the service. 

I use this role in the form of this PR and for me it works well.

**Comment by 0x0I** [comment]
> > > Hi, any news about having the possibility to enable created services?
> > 
> > 
> > _See comment below._
> > Feels like rendering of _unit_ files (the responsibility of this role) and actual orchestration of the resource described by that file (responsibility of e.g. Ansible's **service** or **systemd** module) should be separate/independent. Not to mention that an orchestration solution is already well implemented (see previously mentioned modules lol).
> 
> My use case is somewhat different, I include this role in the playbook and the host_var contains the `unit_config`. Thats it, there is no role, no logic at all. I only want to add the unit files and activate the service.
> 
> I use this role in the form of this PR and for me it works well.

Makes sense and I'm open to others' thoughts but seems like adding unit control at this level conflates two actions (configuration and operation) which could probably be expressed more succinctly with an additional `include_tasks` following the `include_role` call in your playbook (with perhaps a list of units to enable, start, etc as an input var).

**Comment by jstenvall-ducksify** [comment]
> > > > Hi, any news about having the possibility to enable created services?
> > > 
> > > 
> > > _See comment below._
> > > Feels like rendering of _unit_ files (the responsibility of this role) and actual orchestration of the resource described by that file (responsibility of e.g. Ansible's **service** or **systemd** module) should be separate/independent. Not to mention that an orchestration solution is already well implemented (see previously mentioned modules lol).
> > 
> > 
> > My use case is somewhat different, I include this role in the playbook and the host_var contains the `unit_config`. Thats it, there is no role, no logic at all. I only want to add the unit files and activate the service.
> > I use this role in the form of this PR and for me it works well.
> 
> Makes sense and I'm open to others' thoughts but seems like adding unit control at this level conflates two actions (configuration and operation) which could probably be expressed more succinctly with an additional `include_tasks` following the `include_role` call in your playbook (with perhaps a list of units to enable, start, etc as an input var).

I totally understand your point of view. So, what do you suggest? I have no issue with scrapping this PR. If thats the case, I'm most likely going to fork this role and make my own changes in my tree.

Thanks for your time and valuable insights.

**Comment by 0x0I** [comment]
> > > > > Hi, any news about having the possibility to enable created services?
> > > > 
> > > > 
> > > > _See comment below._
> > > > Feels like rendering of _unit_ files (the responsibility of this role) and actual orchestration of the resource described by that file (responsibility of e.g. Ansible's **service** or **systemd** module) should be separate/independent. Not to mention that an orchestration solution is already well implemented (see previously mentioned modules lol).
> > > 
> > > 
> > > My use case is somewhat different, I include this role in the playbook and the host_var contains the `unit_config`. Thats it, there is no role, no logic at all. I only want to add the unit files and activate the service.
> > > I use this role in the form of this PR and for me it works well.
> > 
> > 
> > Makes sense and I'm open to others' thoughts but seems like adding unit control at this level conflates two actions (configuration and operation) which could probably be expressed more succinctly with an additional `include_tasks` following the `include_role` call in your playbook (with perhaps a list of units to enable, start, etc as an input var).
> 
> I totally understand your point of view. So, what do you suggest? I have no issue with scrapping this PR. If thats the case, I'm most likely going to fork this role and make my own changes in my tree.
> 
> Thanks for your time and valuable insights.

So I'm actually curious what your setup looks like. I mean, I support the fork and hate for you to scrap the PR tho feels like you could accomplish your use-case in a pretty painless way.

**Comment by rook59** [comment]
I am also interested in enabling and starting a systemd service with this role.

> So I'm actually curious what your setup looks like

group / host_vars which contain unit_config variable
playbook yml which contains import_role


To enable / start the systemd services I would need yet another task in each playbook which uses this role to loop unit_config and enable / start all the services I want to start.

**Comment by 0x0I** [comment]
FYI a merged a change which should cover the enabling/setting state functionality - https://github.com/0x0I/ansible-role-systemd/releases/tag/v0.3.0

**Comment by jstenvall-ducksify** [comment]
Thanks a lot for this code change. It is much more elegant than my PR. I'll close it with no further actions needed.

Thanks again!

**Comment by goldpanic** [comment]
Thank you very much 👌

**Pull request closed without merging**